go/pkg/cas/upload.go

```diff
@@ -39,6 +65,361 @@ type DigestStat struct {
 
 // Upload uploads all inputs. It exits when inputC is closed or ctx is canceled.
 func (c *Client) Upload(ctx context.Context, inputC <-chan *UploadInput) (stats *TransferStats, err error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	// Do not exit until all sub-goroutines exit, to prevent goroutine leaks.
+	defer eg.Wait()
+
+	u := &uploader{
+		Client: c,
+		eg:     eg,
+
+		fsSem: semaphore.NewWeighted(int64(c.FSConcurrency)),
+	}
+
+	// Start processing input.
+	eg.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case in, ok := <-inputC:
+				if !ok {
+					return nil
+				}
+				u.eg.Go(func() error {
+					return errors.Wrapf(u.startProcessing(ctx, in), "%q", in.Path)
+				})
+			}
+		}
+	})
+	return &u.stats, eg.Wait()
+}
+
+// uploader implements a concurrent multi-stage pipeline to read blobs from the
+// file system, check their presence on the server and then upload if necessary.
+// Common blobs are deduplicated.
+//
+// uploader.eg is used to schedule work, while concurrency of individual
+// expensive operations is controlled via separate semaphores.
+//
+// Special care is taken for large files: they are read sequentially, opened
+// only once per file, and read with large IO size.
+type uploader struct {
+	*Client
+	eg    *errgroup.Group
+	stats TransferStats
+
+	// already-processed files.
+	// Types fsCacheKey and fsCacheValue are used for keys/values respectively.
+	fsCache sync.Map
```

Comment:
https://golang.org/pkg/sync/#Map says
```
The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.
```

Can we start from map with mutext?

Reply:
switched to mutex and add a TODO to consider sync.Mutex

Reply:
"The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow ...". I think syncMap is appropriate for this use case, but in general this seems very close to singleflightcache. Would it make sense to use the same package for the same functionality instead of having different implementations? We can address performance concerns in the singleflightcache itself and just use it here.

Reply:
@ramymedhat just to see if we are thinking about the same thing: do you mean that reusing a common functionality could save us 3 lines of code?

```go
	entryRaw, _ := u.fsCache.LoadOrStore(cacheKey, &fsCacheValue{})
	e := entryRaw.(*fsCacheValue)
	e.compute.Do(func() {
```

Is this over-generalization?

Also, FWIU the only usage of `singleflightcache` is for CAS, so if we switch to this package, do we still need `singleflightpackage`?

Reply:
My point is more about reusing the common lib to make sure whatever optimizations we do apply here and elsewhere. For instance, I think here you're doing one less lock which would be beneficial in the general implementation. Also with a central cache we can track its size and memory consumption, get cache hit rates and other stats.

Reply:
> Also, FWIU the only usage of singleflightcache is for CAS, so if we switch to this package, do we still need singleflightpackage?

We use the cache for other use cases besides file digest. We also use the file metadata cache outside of the SDK.

Reply:
+1 to what Ramy said. This can be useful outside of the SDK (to lookup metadata/digest checking). Enhancing `SingleFlightCache` (which likely can perform better) or eventually (in a later PR?) exporting this both sound good to me.

Reply:
> We also use the file metadata cache outside of the SDK.

Thanks, this didn't occur to me.

I am happy to improve singleflightcache in a separate CL (otherwise seems to be a scope creep). I just tried, but the semantics is not the same: `singleflightcache` avoids caching the error (if `valFn` errors out, then the value isn't stored), so the same trick doesn't work. `uploader` here doesn't need that, and thus can be simpler (just one map, with `once`s) and more efficient. Ideas?

Reply:
I *think* those semantics aren't necessary. We might just need the stat metadata saved, which can be done at a later PR. @ramymedhat ?

Reply:
Yeah I think we don't need that. Ok to defer to a later PR though.

Reply:
I sent out https://github.com/bazelbuild/remote-apis-sdks/pull/303, but I couldn't figure out how to rebase this PR on top of the new one (something that I do in gerrit all the time). So I am going to have to merge this PR, and then update https://github.com/bazelbuild/remote-apis-sdks/pull/303 to use `singleflightcache` in `cas` package.